internal/controllers/objectsetphases/phase_reconciler_test.go

```diff
@@ -0,0 +1,733 @@
+package objectsetphases
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"package-operator.run/apis/core/v1alpha1"
+	corev1alpha1 "package-operator.run/apis/core/v1alpha1"
+	"package-operator.run/package-operator/internal/controllers"
+	"package-operator.run/package-operator/internal/testutil"
+)
+
+func TestPhaseReconciler_TeardownPhase(t *testing.T) {
+	t.Run("already gone", func(t *testing.T) {
+		dynamicCache := &dynamicCacheMock{}
+		ownerStrategy := &ownerStrategyMock{}
+		r := &PhaseReconciler{
+			dynamicCache:  dynamicCache,
+			ownerStrategy: ownerStrategy,
+		}
+		owner := &unstructured.Unstructured{}
+
+		ownerStrategy.
+			On("SetControllerReference", mock.Anything, mock.Anything, mock.Anything).
+			Return(nil)
+
+		dynamicCache.
+			On("Get", mock.Anything, mock.Anything, mock.Anything, mock.Anything).
+			Return(errors.NewNotFound(schema.GroupResource{}, ""))
+
+		ctx := context.Background()
+		done, err := r.TeardownPhase(ctx, owner, corev1alpha1.ObjectSetTemplatePhase{
+			Objects: []corev1alpha1.ObjectSetObject{
+				{
+					Object: runtime.RawExtension{},
+				},
+			},
+		})
+		require.NoError(t, err)
+		assert.True(t, done)
+	})
+
+	t.Run("delete confirms", func(t *testing.T) {
+		testClient := testutil.NewClient()
+		dynamicCache := &dynamicCacheMock{}
+		ownerStrategy := &ownerStrategyMock{}
+		r := &PhaseReconciler{
+			writer:        testClient,
+			dynamicCache:  dynamicCache,
+			ownerStrategy: ownerStrategy,
+		}
+		owner := &unstructured.Unstructured{}
+
+		ownerStrategy.
+			On("SetControllerReference", mock.Anything, mock.Anything, mock.Anything).
+			Return(nil)
+
+		currentObj := &unstructured.Unstructured{}
+		dynamicCache.
+			On("Get", mock.Anything, mock.Anything, mock.Anything, mock.Anything).
+			Run(func(args mock.Arguments) {
+				out := args.Get(2).(*unstructured.Unstructured)
+				*out = *currentObj
+			}).
+			Return(nil)
+
+		ownerStrategy.
+			On("IsController", owner, currentObj).
+			Return(true)
+
+		testClient.
+			On("Delete", mock.Anything, mock.Anything, mock.Anything).
+			Return(errors.NewNotFound(schema.GroupResource{}, ""))
+
+		ctx := context.Background()
+		done, err := r.TeardownPhase(ctx, owner, corev1alpha1.ObjectSetTemplatePhase{
+			Objects: []corev1alpha1.ObjectSetObject{
+				{
+					Object: runtime.RawExtension{},
+				},
+			},
+		})
+		require.NoError(t, err)
+		assert.True(t, done)
+
+		// It's super important that we don't check ownership on desiredObj on accident, because that will always return true.
```

Comment:
```suggestion
		// It's super important that we check IsController, not IsOwner, because IsOwner will always return true.
```

Reply:
I am more interested in the parameters given to `IsController`. It has to be called with `currentObj` and not `desiredObj`.
Trying to make the comment clearer... :)

Reply:
Ohhhhhhhhhhhhh now I get it :D

Reply:
I think I was confused because because we are checking if it is a controller, not directly checking ownership, right?